### PR TITLE
Crosschain delegatable resolver with subname

### PR DESCRIPTION
**Pull request by makoto**
Changes

- Move `targets` to private and reintroduce `getTarget(encodedname, offset)` that can find target address for subnames set by parents.
- Introduce `resolver(encodedname, calldata)` and move `addr/text/contenthash` private
- Add test  to resolve ETH Address for subname

### CODE REVIEW

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 1]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
     }
 
     /**
-     * Returns the address associated with an ENS node.
-     * @param node The ENS node to query.
-     * @return The associated address.
+     * Query the target address that is verifiable against the name
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:69]:
```suggestion
     * @dev Returns the L2 target address that can answer queries for `name`.
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 2]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
...
+     * @param name DNS encoded ENS name to query
+     * @param offset The offset of the label to query recursively. Start from the 0 position and kepp adding the length of each label as it traverse. The function exits when len is 0.
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:71]:
```suggestion
     * @param offset The offset of the label to query recursively.
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 3]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
...
+        node = bytes32(0);
+        if (len > 0) {
+            bytes32 label = name.keccak(offset + 1, len);
+            (node, target) = getTarget(
+                name,
+                offset + len + 1
+            );
+            node = keccak256(abi.encodePacked(node, label));
+            if(targets[node] != address(0)){
+              target = targets[node];
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:89]: You can return here.

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 4]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
...
+            }
+        } else {
+            return (
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:92]: Returning 0,0 would be clearer.

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 5]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
...
+     * @dev Resolve and verify a record stored in l2 target address. It supports subname by fetching target recursively to the nearlest parent.
+     * @param name DNS encoded ENS name to query
+     * @param data The actual calldata
+     * @return result result of the call
      */
-    function addr(bytes32 node) public view returns (address) {
-        EVMFetcher.newFetchRequest(verifier, targets[node])
+    function resolve(bytes calldata name, bytes calldata data) external view returns (bytes memory result) {
+        (, address target) = getTarget(name, 0);
+        bytes4 seletor = bytes4(data);
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:108]:
```suggestion
        bytes4 selector = bytes4(data);
```

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 6]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
...
+        if (seletor == 0x3b3b57de) {
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:109]: These should be constants - ideally, `SomeInterface.func.selector`.

File: crosschain-resolver/contracts/L1Resolver.sol [hunk 7]
```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
...
+            (bytes32 node, string memory key) = abi.decode(data[4:], (bytes32, string));
+            return bytes(text(node, key, target));
+        }
+        if (bytes4(data) == 0xbc1c58d1) {
+            (bytes32 node) = abi.decode(data[4:], (bytes32));
+            return contenthash(node, target);
+        }
+    }
+
+    function addr(bytes32 node, address target) private view returns (bytes memory) {
```
Comment (Arachnid) [inline crosschain-resolver/contracts/L1Resolver.sol:128]: These functions should probably be named differently to avoid confusion.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: crosschain-resolver/contracts/L1Resolver.sol (hunk 6 above)
Comment (makoto) [reply crosschain-resolver/contracts/L1Resolver.sol]: Couldn't figure out how to do that for functions with same name and different arguments (eg: `addr(bytes32,uint)` so used `bytes4(keccak256("addr(bytes32,uint256)"))` instead.

**Review by Arachnid — APPROVED** [review]

**Pull request merged**